mulog-core/src/com/brunobonacci/mulog/publisher.clj

```diff
@@ -100,18 +100,28 @@
 
 
 
+(defn- make-parent-dirs [f]
+  (condp instance? f
+    String
+    (make-parent-dirs (io/file f))
+
+    java.io.File
+    (when-let [path (.getParentFile ^java.io.File f)]
+      (.mkdirs path))
+
+    nil))
+
+
+
 (defn simple-file-publisher
   [{:keys [filename transform] :as config}]
   {:pre [filename]}
-  (let [filename (io/file filename)]
-    ;; make parent dirs
-    (when-let [path (.getParentFile filename)]
-      (.mkdirs path))
-    (SimpleFilePublisher.
-      config
-      (io/writer filename :append true)
-      (rb/agent-buffer 10000)
-      (or transform identity))))
+  (make-parent-dirs filename)
```

Comment:
I see, so for  Hadoop `FSDataOutputStream` you would construct the `FSDataOutputStream` and pass it in place of the `filename`.

I thought you would extend the `clojure.java.io/IOFactory` and let `io/output-stream` do its job.

so I guess the call to `io/make-parents` needs to be guarded for the right type

``` Clojure
(when (or (string? filename) (instance? java.io.File filename))
  (io/make-parents filename))
```

Reply:
That makes sense, I wasn't aware of `io/make-parents`.
I've included a check so that it will only try to make parents on types that it makes sense for.
I'm pretty sure it won't work on HDFS destinations, but I think that's fine - in many cases (e.g. if it's pointing to S3) it won't need the parent to be created first anyway, and if it does I think it's fine to leave that as the users responsibility.

Reply:
yes, but in that case, `(io/make-parents "s3://some/path.txt")` won't work, it will probably create a local directory called `s3:`.

Reply:
Oh, I see what you mean... yes I just tried it and it created `s3:`/`some`

Reply:
I totally agree.

I just don't like the following:

``` Clojure
  (when (some #(instance? % filename)
              (remove nil? (extenders io/Coercions)))
    (io/make-parents filename))
``` 

To me, this is not clear what it is happening here, and it could cause unpredictable behaviour as the order of the `extenders` returns is not predictable (depends on code loading and hash function).

I rather have an explicit test for string and files, like it was it your initial PR on in the comment above.

``` Clojure
(when (or (string? filename) (instance? java.io.File filename))
  (io/make-parents filename))
```

What do you think?

Reply:
I was trying to make it so that this would work with anything that could be coerced to a File, whether a String, a java File or anything else. But I agree that maybe this is trying to be too clever, and just testing for String or File as before is clearer and simpler. I'll change it back.